dev/com.ibm.ws.org.apache.cxf.cxf.rt.frontend.jaxrs.3.2/src/org/apache/cxf/jaxrs/provider/ProviderFactory.java

```diff
@@ -830,36 +860,104 @@ void injectContextProxiesIntoProvider(ProviderInfo<?> pi, Application app) {
         }
     }
 
-    /*
-     * sorts the available providers according to the media types they declare
-     * support for. Sorting of media types follows the general rule: x/y < * x < *,
-     * i.e. a provider that explicitly lists a media types is sorted before a
-     * provider that lists *. Quality parameter values are also used such that
-     * x/y;q=1.0 < x/y;q=0.7.
-     */
-    private void sortReaders() {
+    //Liberty code change start
+    private void addAndSortReaders(List<ProviderInfo<MessageBodyReader<?>>> newReaders) {
+        Comparator<ProviderInfo<MessageBodyReader<?>>> comparator = null;
         if (!customComparatorAvailable(MessageBodyReader.class)) {
-            messageReaders.sort(new MessageBodyReaderComparator(readerMediaTypesMap));
-        } else {
-            doCustomSort(messageReaders);
+            comparator = new MessageBodyReaderComparator(readerMediaTypesMap);
         }
+
+        messageReaders.addAndSortProviders(newReaders, comparator);
     }
 
-    private <T> void sortWriters() {
+    private void addAndSortWriters(List<ProviderInfo<MessageBodyWriter<?>>> newWriters) {
+        Comparator<ProviderInfo<MessageBodyWriter<?>>> comparator = null;
         if (!customComparatorAvailable(MessageBodyWriter.class)) {
-            messageWriters.sort(new MessageBodyWriterComparator(writerMediaTypesMap));
-        } else {
-            doCustomSort(messageWriters);
+            comparator = new MessageBodyWriterComparator(writerMediaTypesMap);
         }
-        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-            StringBuilder msg = new StringBuilder("sortWriters - sorted list:");
-            for (int i = 0; i < messageWriters.size(); i++) {
-                msg.append(" (" + i + ") " + messageWriters.get(i).getProvider());
+
+        messageWriters.addAndSortProviders(newWriters, comparator);
+    }
+
+    protected class AtomicReferenceProviderList<T> implements Iterable<ProviderInfo<T>> {
+        private final AtomicReference<List<ProviderInfo<T>>> referent;
+
+        public AtomicReferenceProviderList() {
+            referent = new AtomicReference<>(Collections.emptyList());
+        }
+
+        public AtomicReferenceProviderList(List<ProviderInfo<T>> list) {
```

Comment:
same comment as earlier - is this constructor ever called?  if not, I'd suggest removing it, since it introduces the possibility of a NPE in `addAndSortProviders(...)`